src/github.com/t3chguy/riot-static/riot-static.go

```diff
@@ -0,0 +1,234 @@
+		"NumMembers": length,
+	})
+	data.RUnlock()
+}
+
+func GetPublicRoomPowerLevels(c *gin.Context) { // ~/room/:roomId/power_levels
+	roomId := c.Param("roomId")
+
+	data.RLock()
+	c.HTML(http.StatusOK, "power_levels.html", gin.H{
+		"PowerLevels": data.Rooms[roomId].PowerLevels,
+	})
+	data.RUnlock()
+}
+
+func GetPublicRoomMember(c *gin.Context) { // ~/room/:roomId/members/:mxid
+	roomId := c.Param("roomId")
+	mxid := c.Param("mxid")
+
+	data.RLock()
+	if memberInfo := data.Rooms[roomId].MemberInfo[mxid]; memberInfo != nil {
+		c.HTML(http.StatusOK, "member_info.html", gin.H{
+			"RoomID":     roomId,
+			"MXID":       mxid,
+			"MemberInfo": memberInfo,
+		})
+	} else {
+		c.AbortWithStatus(http.StatusNotFound)
+	}
+	data.RUnlock()
+}
+
+var data = struct {
+	resync.Once
+	sync.RWMutex
+	NumRooms int
+	Ordered  []*Room
+	Rooms    map[string]*Room
+}{}
+
+func LoadPublicRooms() {
+	// @TODO: fix this.
+	data.Lock()
+	fmt.Println("Loading public publicRooms")
+	resp, err := cli.PublicRooms(0, "", "")
+
+	if err == nil {
+		b := []*Room{}
+		c := map[string]*Room{}
+
+		// filter on actually WorldReadable publicRooms
+		for _, x := range resp.Chunk {
+			if x.WorldReadable {
+				var room *Room
+				if data.Rooms[x.RoomId] != nil {
+					room = data.Rooms[x.RoomId]
+				} else {
+					room = NewRoom(x)
+				}
+				b = append(b, room)
+				c[x.RoomId] = room
+			}
+		}
+
+		//data.Lock()
+		data.Rooms = c
+		data.NumRooms = len(b)
+		// copy order so we don't encounter slice hell
+		data.Ordered = make([]*Room, data.NumRooms)
+		copy(data.Ordered, b)
+
+		data.Unlock()
+	}
+
+	if err != nil {
+		panic(err)
+	}
+}
+
+func main() {
+	setupCli()
+
+	router := gin.Default()
+	router.SetHTMLTemplate(tpl)
+	router.Static("/assets", "./assets")
+
+	router.Use(func(c *gin.Context) {
+		data.Once.Do(LoadPublicRooms)
+	})
+
+	router.GET("/", GetPublicRoomsList)
+
+	roomRouter := router.Group("/room/")
+	{
+		roomRouter.Use(func(c *gin.Context) {
+			roomId := c.Param("roomId")
+			if data.Rooms[roomId] == nil {
+				c.String(http.StatusNotFound, "Room Not Found")
+				c.Abort()
+			} else {
+				// Start of debug code
+				if _, exists := c.GetQuery("clear"); exists {
+					data.Rooms[roomId].Once.Reset()
+				}
+				// End of debug code
+				data.Rooms[roomId].Fetch()
+				c.Next()
+			}
+		})
+
+		roomRouter.GET("/:roomId/", func(c *gin.Context) {
+			c.Redirect(http.StatusTemporaryRedirect, "chat")
+		})
+		roomRouter.GET("/:roomId/chat", GetPublicRoom)
+		roomRouter.GET("/:roomId/servers", GetPublicRoomServers)
+		roomRouter.GET("/:roomId/members", GetPublicRoomMembers)
+		roomRouter.GET("/:roomId/members/:mxid", GetPublicRoomMember)
+		roomRouter.GET("/:roomId/power_levels", GetPublicRoomPowerLevels)
+	}
+
+	router.GET("/clear", func(c *gin.Context) {
+		data.Once.Reset()
+		for _, room := range data.Rooms {
+			room.Once.Reset()
+		}
+	})
+
+	port := os.Getenv("PORT")
+	if port == "" {
+		port = "8000"
+	}
+
+	router.Run(":" + port)
```

Comment:
This uses the default HTTP server which has no timeouts implemented.
- https://godoc.org/github.com/gin-gonic/gin#Engine.Run
- https://golang.org/pkg/net/http/#ListenAndServe
- https://golang.org/src/net/http/server.go?s=83696:83751#L2775
- https://golang.org/pkg/net/http/#Server

See https://blog.gopheracademy.com/advent-2016/exposing-go-on-the-internet/ for a very thorough review of this. You need to be able to give your own `http.Server` to gin.